src/Agda2Hs/Compile/Record.hs

```diff
@@ -168,16 +170,30 @@ compileRecord target def = do
       let conDecl = Hs.QualConDecl () Nothing Nothing $ Hs.RecDecl () cName fieldDecls
       return $ Hs.DataDecl () don Nothing hd [conDecl] ds
 
+checkUnboxPragma :: Definition -> C ()
+checkUnboxPragma def = do
+  let Record{..} = theDef def
 
--- | Check if record can be defined as unboxed.
-checkUnboxPragma :: Defn -> C ()
-checkUnboxPragma def
-  | Record{recFields, recMutual} <- def
-  , length (filter keepArg recFields) == 1
-  -- , not (recRecursive def)
-  --     can be used again after agda 2.6.4.2 is released
-  --     see: agda/agda#7042
-  , all null recMutual -- see agda/agda#7042
-  = return ()
-  | otherwise
-  = genericError "An unboxed type must be a non-recursive record type with exactly one non-erased field."
+  -- recRecursive can be used again after agda 2.6.4.2 is released
+  -- see agda/agda#7042
+  unless (all null recMutual) $ genericDocError
+    =<< text "Unboxed record" <+> prettyTCM (defName def) 
+    <+> text "cannot be recursive"
+
+  TelV tel _ <- telViewUpTo recPars (defType def)
+  addContext tel $ do
+    pars <- getContextArgs
+    let fieldTel = recTel `apply` pars
+    fields <- nonErasedFields fieldTel
+    unless (length fields == 1) $ genericDocError
+      =<< text "Unboxed record" <+> prettyTCM (defName def)
+      <+> text "should have exactly one non-erased field"
+
+  where
+    nonErasedFields :: Telescope -> C [String]
+    nonErasedFields EmptyTel = return []
+    nonErasedFields (ExtendTel a tel) = compileDom a >>= \case
+      DODropped  -> underAbstraction a tel nonErasedFields
+      DOType -> genericDocError =<< text "Type field in unboxed record not supported"
+      DOInstance -> genericDocError =<< text "Instance field in unboxed record not supported"
```

Comment:
Do we not? I assumed since #284 instance search can be used to generate Haskell terms, wouldn't this also work fine for instance fields in unboxed records? Don't know if it'd ever useful to define an unboxed record so that the field is always resolved by instance search, but anyways.

Unless instance arguments for non-class records already produce a `DOTerm` output? Would make sense.

Reply:
I did not break any test cases so the feature was never really supported :)

Reply:
> Unless instance arguments for non-class records already produce a DOTerm output? Would make sense.

This is indeed the case.